### PR TITLE
Prescriptions Page for Patients (Home Page)

### PR DESCRIPTION
**Pull request by JJwilkin**

### CODE REVIEW

File: pages/api/locations/read.ts
Comment (rajwanialiya) [inline pages/api/locations/read.ts:25]: If we can already use the perscriptions /read endpoint to retrieve perscriptions by locationId, how come you want to include it here too?

File: pages/home.tsx
Comment (rajwanialiya) [inline pages/home.tsx]: id

File: pages/home.tsx
Comment (rajwanialiya) [inline pages/home.tsx]: remove

File: helpers/user-details.ts
Comment (SammyRobensParadise) [inline helpers/user-details.ts]: this should be passed as the query object on the homepage
```suggestion
```

File: pages/api/locations/read.ts
Comment (SammyRobensParadise) [inline pages/api/locations/read.ts:15]: is there a reason why we are just sending the body? and not data? generally, we want to pass in a data object as well. If this is being changed, make sure to update all the places where this is called as well.

File: pages/api/locations/read.ts
Comment (SammyRobensParadise) [reply pages/api/locations/read.ts:25]: the better way to do this is to get this information on the `user-details` hook by writing a query for initial data fetching, similar to how to was done for other user attributes. See the `team.tsx` page as an example

File: pages/api/locations/read.ts
Comment (SammyRobensParadise) [reply pages/api/locations/read.ts:15]: alternatively, and maybe a better idea is to make this file called `pages/api/locations/[id].ts` and then get the id using `req.query` this way we can actually define the endpoint with a `GET` request appropriately

File: pages/api/prescriptions/read.ts
Comment (SammyRobensParadise) [inline pages/api/prescriptions/read.ts]: add type definitions here so we know what we are expecting

File: pages/api/prescriptions/read.ts
Comment (SammyRobensParadise) [inline pages/api/prescriptions/read.ts:9]: It may be worth making this a `GET` endpoint as per my previous comments. refer to the [next.js API documentation](https://nextjs.org/docs/api-routes/dynamic-api-routes)

File: pages/api/prescriptions/read.ts
Comment (SammyRobensParadise) [inline pages/api/prescriptions/read.ts:36]: these two calls should really be two separate endpoints. one should be on the user route since we are getting the "users" or the "patients" prescriptions. the `api/prescriptions` endpoints should be when we want to get all prescriptions but like to get a specific users I think It could be

`GET /api/patient/[id]/presciptions`

File: pages/api/prescriptions/read.ts
Comment (SammyRobensParadise) [reply pages/api/prescriptions/read.ts:9]: Refer to later comment about potentially re-structuring endpoints

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]: All of this code can be restructured. in the `getServerSideProps` you can add a `query` to get the patient Info and staff info see this example on the teams page https://github.com/pharmaboxfydp/pharmabox/blob/9ad02464ce50ab98909831a6b5d64b2fcfdb400a/pages/team.tsx#L71

Comment (SammyRobensParadise) [reply pages/home.tsx]: Then you are going to want to restructure this into a react hook using `swr`
see this as an example:
 https://github.com/pharmaboxfydp/pharmabox/blob/main/hooks/team.ts#L29
and then it can be used like so https://github.com/pharmaboxfydp/pharmabox/blob/9ad02464ce50ab98909831a6b5d64b2fcfdb400a/components/TeamMembersTable.tsx#L26

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]:
```suggestion
                key="qrCode"
```

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]:
```suggestion
            <Box pad="xlarge">
```

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]: Add a loader. you can use `<Spinner/> or `<Skeleton/>` see this example

https://github.com/pharmaboxfydp/pharmabox/blob/9ad02464ce50ab98909831a6b5d64b2fcfdb400a/components/TeamMembersTable.tsx#L50

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]:
```suggestion
                    margin="large"
```

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]:
```suggestion
```

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx]:
```suggestion
```

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx:118]: This whole block is a pretty large conditional render based on the user role. I would suggest splitting it up into separate components. Refer to the settings page as an example 
https://github.com/pharmaboxfydp/pharmabox/blob/9ad02464ce50ab98909831a6b5d64b2fcfdb400a/pages/settings/%5B%5B...index%5D%5D.tsx#L52

File: types/types.ts
Comment (SammyRobensParadise) [inline types/types.ts]:
```suggestion
  Prescriptions: Prescription[]
```

File: types/types.ts
Comment (SammyRobensParadise) [inline types/types.ts]:
```suggestion
  LockerBoxes: LockerBox[]
```

**Review by SammyRobensParadise — CHANGES_REQUESTED** [review]
Looks good! a couple of things! :)

File: pages/api/locations/read.ts
Comment (JJwilkin) [reply pages/api/locations/read.ts:15]: currently we're not passing in any additional information in the body so adding a data field doesn't provide any functional difference. Although this could be helpful in the future if we want to pass other information in the body but not the data object. I'm not sure what that use case would be though

File: components/PatientPrescriptions.tsx
Comment (SammyRobensParadise) [inline components/PatientPrescriptions.tsx:22]: nit pick but you can enforce function definition here
```suggestion
  setQrCode: (pickupCode: string) => void
```

File: components/PatientPrescriptions.tsx
Comment (SammyRobensParadise) [inline components/PatientPrescriptions.tsx:23]:
```suggestion
  setShowQrCode: (state:boolean) => void
```

File: components/QrCode.tsx
Comment (SammyRobensParadise) [inline components/QrCode.tsx:14]:
```suggestion
const QRCodeModal = ({ qrCode, onClose }: { qrCode: string; onClose: (state:boolean) => void }) => {
```

File: components/QrCode.tsx
Comment (SammyRobensParadise) [inline components/QrCode.tsx:32]: You can use the heading elements from `grommet` https://v2.grommet.io/heading

File: components/QrCode.tsx
Comment (SammyRobensParadise) [inline components/QrCode.tsx:40]:
```suggestion
              margin='large'
```

File: components/QrCode.tsx
Comment (SammyRobensParadise) [inline components/QrCode.tsx:22]:
```suggestion
      <Box pad='xlarge'>
```

File: pages/api/locations/read.ts
Comment (SammyRobensParadise) [reply pages/api/locations/read.ts:15]: makes sense

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx:84]: You might want to put this in a hook and call it `usePrescriptions` and `useLocation` and then use it in the patient prescriptions component. It will make this a lot easier to read I think you can have it return some nice things imagine using it like
just create and call this hook in `components/PatientPrescriptions.tsx` and then all you need to pass in is the `user`
```tsx
{user.role ===Role.Patient && <PatientPrescriptions user={user}>}
```
```tsx
const { awaitingPickup, pickupCompleted, isLoading, isError } = usePrescriptions(user)
```

then in `components/PharmacyLocation.tsx` call a hook that looks like this:
```tsx
const { location, isLoading, isError } = useLocation(user)
```
```tsx
{user.role ===Role.Staff && <PharmacyLocationView user={user}>}
```

this change will make sure that we are only ever creating stateful elements that intend to use, reduces the number of conditional renders, and mean that you don't have to store any sort of state because we are getting it from the hooks. LMK if this makes sense. I am not gonna block the PR from merging, so up to you if you want to make this change before merging it.

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx:117]: we have `enums for these` defined in `/types/types.ts` ! 
```suggestion
          {user.role === Role.Patient && (
            <PatientPrescriptions
              prescriptions={prescriptions}
              previousPrescriptions={previousPrescriptions}
              setQrCode={setQrCode}
              setShowQrCode={setShowQrCode}
            />}
            {role === Role.Staff &&<PharmacyLocationView pharmacyLocation={pharmacyLocation} /> }
```
removing the ternary operator makes it a little safer since this way we will only load the components if there is an exact match. i.e lets say the user's role is `undefined` for some reason (maybe a failed fetch or a data error), then in this case we would render the Staff view which we wouldn't want to happen. Instead we would want to show nothing, or throw up an error screen.

File: types/types.ts
Comment (SammyRobensParadise) [inline types/types.ts:22]: this works! you can also use the typescript `extends` definer https://www.typescriptlang.org/docs/handbook/interfaces.html#extending-interfaces. 

can sometimes make it easier to see the true nested structure of types

File: pages/home.tsx
Comment (SammyRobensParadise) [inline pages/home.tsx:95]: because you are doing this at the page level. we are blocking access to the rest of the application. If you move the fetching into the components, then you can resolve this "full page load" issue
Screen Shot 2022-11-06 at 1 18 24 AM

**Review by SammyRobensParadise — APPROVED** [review]
I am gonna approve because I don't think anything here is "blocking" in terms of the LFP. there are a couple of "quick wins" that may make it easier and deal with some issues that you can choose to implement or not. I am wondering though if you can add (to this PR or another) to the `seed_test_dbs` script a bunch of prescriptions associated with patients so that we can easily test going forward. What are your thoughts?

**Pull request merged**
